File: leetcode/Search_for_a_Range.java
import java.util.*;

public class Solution {
    public int[] searchRange(int[] A, int target) {
        int a = 0;
        int b = A.length;
        boolean flag = false;
        while (a < b) {
            int m = (a + b) / 2;
            if (A[m] == target) {
                flag = true;
                break;
            } else if (A[m] < target) {
                a = m + 1;
            } else {
                b = m;
            }
        }
        int[] ret = new int[2];
        if (!flag) {
            ret[0] = ret[1] = -1;
            return ret;
        }
        if (A[0] == target) {
            ret[0] = 0;
        } else {
            a = 0;
            b = A.length - 1;
            while (a + 1 < b) {
                int m = (a + b) / 2;
                if (A[m] < target) a = m;
                else b = m;
            }
            ret[0] = b;
        }
        if (A[A.length - 1] == target) {
            ret[1] = A.length - 1;
        } else {
            a = 0;
            b = A.length - 1;
            while (a + 1 < b) {
                int m = (a + b) / 2;
                if (A[m] <= target) a = m;
                else b = m;
            }
            ret[1] = a;
        }
        return ret;
    }
}

File: leetcode/Search_for_a_Range.cpp
class Solution {
    int binarySearch(int A[], int beg, int end, int target, bool accept_equal) {
        if (beg + 1 == end) {
            if (A[beg] > target || (accept_equal && A[beg] == target))
                return beg - 1;
            else
                return beg;
        }
        int mid = (beg + end) / 2;
        if (A[mid] > target || (accept_equal && A[mid] == target))
            return binarySearch(A, beg, mid, target, accept_equal);
        else
            return binarySearch(A, mid, end, target, accept_equal);
    }
public:
    vector<int> searchRange(int A[], int n, int target) {
        int first_pos = binarySearch(A, 0, n, target, true) + 1;
        int last_pos  = binarySearch(A, 0, n, target, false);
        vector<int> range(2, -1);
        if (first_pos <= last_pos) {
            range[0] = first_pos;
            range[1] = last_pos;
        }
        return range;
    }
};
